core/utils/thread_pool.hpp

```diff
@@ -14,40 +15,184 @@
+        -> std::optional<std::tuple<Args...>> {
+      BOOST_ASSERT(handler_);
+      return handler_->reinvoke(std::forward<F>(func),
+                                std::forward<Args>(args)...);
+    }
+
+    bool isInCurrentThread() const {
+      BOOST_ASSERT(handler_);
+      return handler_->isInCurrentThread();
+    }
+
+    ThreadHandler &handler() {
+      BOOST_ASSERT(handler_);
+      return *handler_;
     }
 
    private:
-    std::shared_ptr<boost::asio::io_context> io_context_;
-    boost::asio::executor_work_guard<boost::asio::io_context::executor_type>
-        work_guard_;
+    std::optional<ThreadHandler> handler_;
     std::vector<std::thread> threads_;
   };
 }  // namespace kagome
 
+#define REINVOKE_3(ctx, func, in_1, in_2, in_3, out_1, out_2, out_3) \
+  auto res##func = (ctx).reinvoke(                                   \
+      [wself(weak_from_this())](auto &&x1, auto &&x2, auto &&x3) {   \
+        if (auto self = wself.lock()) {                              \
+          self->func(std::move(x1), std::move(x2), std::move(x3));   \
+        }                                                            \
+      },                                                             \
+      std::move(in_1),                                               \
+      std::move(in_2),                                               \
+      std::move(in_3));                                              \
+  if (!(res##func)) return;                                          \
+  auto &&[func##x1, func##x2, func##x3] = std::move(*(res##func));   \
+  auto && (out_1) = func##x1;                                        \
+  auto && (out_2) = func##x2;                                        \
+  auto && (out_3) = func##x3;
+
+#define REINVOKE_2(ctx, func, in_1, in_2, out_1, out_2)  \
+  auto res##func = (ctx).reinvoke(                       \
+      [wself(weak_from_this())](auto &&x1, auto &&x2) {  \
+        if (auto self = wself.lock()) {                  \
+          self->func(std::move(x1), std::move(x2));      \
+        }                                                \
+      },                                                 \
+      std::move(in_1),                                   \
+      std::move(in_2));                                  \
+  if (!(res##func)) return;                              \
+  auto &&[func##x1, func##x2] = std::move(*(res##func)); \
+  auto && (out_1) = func##x1;                            \
+  auto && (out_2) = func##x2;
+
+#define REINVOKE_1(ctx, func, in, out)      \
+  auto res##func = (ctx).reinvoke(          \
+      [wself(weak_from_this())](auto &&x) { \
+        if (auto self = wself.lock()) {     \
+          self->func(std::move(x));         \
+        }                                   \
+      },                                    \
+      std::move(in));                       \
+  if (!(res##func)) return;                 \
+  auto && (out) = std::get<0>(std::move(*(res##func)));
+
```

Comment:
Why `Foo::foo() { REINVOKE(io, foo); ... }` instead of `Foo::foo() { io.post([] { ... }); }`